File: leetcode/easy/232_用栈实现队列/232_用栈实现队列.go
type MyQueue struct {
	stIn  []int // 输入栈
	stOut []int // 输出栈
}

func Constructor() MyQueue {
	return MyQueue{
		stIn:  []int{},
		stOut: []int{},
	}
}

// 辅助函数：将 stIn 元素转移到 stOut
func (q *MyQueue) transfer() {
	if len(q.stOut) == 0 {
		for len(q.stIn) > 0 {
			q.stOut = append(q.stOut, q.stIn[len(q.stIn)-1])
			q.stIn = q.stIn[:len(q.stIn)-1]
		}
	}
}

func (this *MyQueue) Push(x int) {
	this.stIn = append(this.stIn, x)
}

func (this *MyQueue) Pop() int {
	this.transfer()
	res := this.stOut[len(this.stOut)-1]
	this.stOut = this.stOut[:len(this.stOut)-1]
	return res
}

func (this *MyQueue) Peek() int {
	this.transfer()
	return this.stOut[len(this.stOut)-1]
}

func (this *MyQueue) Empty() bool {
	return len(this.stIn) == 0 && len(this.stOut) == 0
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Peek();
 * param_4 := obj.Empty();
 */

File: leetcode/easy/232_用栈实现队列/232_用栈实现队列.cpp
class MyQueue {
private:
    stack<int> stIn;   //输入栈
    stack<int> stOut;  //输出栈

    //辅助函数：将 stIn 元素转移到 stOut
    void transfer() {
        if (stOut.empty()) {
            while (!stIn.empty()) {
                stOut.push(stIn.top());
                stIn.pop();
            }
        }
    }

public:
    void push(int x) {
        stIn.push(x);
    }

    int pop() {
        transfer();
        int res = stOut.top();
        stOut.pop();
        return res;
    }

    int peek() {
        transfer();
        return stOut.top();
    }

    bool empty() {
        return stIn.empty() && stOut.empty();
    }
};



/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */